### PR TITLE
SONARJAVA-1667 use type arguments when resolving constructors

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java [hunk 1]
```diff
@@ -554,8 +555,10 @@ public void visitNewClass(NewClassTree tree) {
     if (newClassTreeImpl.isTypeSet()) {
       return;
     }
+    List<JavaType> typeArgumentsTypes = ImmutableList.of();
     if (tree.typeArguments() != null) {
       resolveAs((List<Tree>) tree.typeArguments(), JavaSymbol.TYP);
+      typeArgumentsTypes = tree.typeArguments().stream().map(typeArgument -> getType(typeArgument)).collect(Collectors.toList());
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)
MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)
MINOR Replace this lambda with a method reference. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1612)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: java-frontend/src/test/files/sym/ConstructorWithTypeArguments.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    new<B>MyClass(b);
+    new<C>MyClass(c);
+  }
+}
+
+interface I {}
+interface J {}
+
+class B implements I {}
+class C implements I, J {}
```
Comment (benzonico) [inline java-frontend/src/test/files/sym/ConstructorWithTypeArguments.java]: missing new line

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java (hunk 1 above)
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: last one should be fixed.

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java [hunk 3]
```diff
@@ -554,8 +555,10 @@ public void visitNewClass(NewClassTree tree) {
     if (newClassTreeImpl.isTypeSet()) {
       return;
     }
+    List<JavaType> typeArgumentsTypes = ImmutableList.of();
     if (tree.typeArguments() != null) {
       resolveAs((List<Tree>) tree.typeArguments(), JavaSymbol.TYP);
+      typeArgumentsTypes = tree.typeArguments().stream().map(this::getType).collect(Collectors.toList());
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)
MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Pull request merged**